### ISSUE TITLE
The semantic of the event in logstash (Text Encoding)

### ISSUE THREAD

**Issue by purbon** [elastic/logstash#5573]
While working on https://github.com/logstash-plugins/logstash-input-jdbc/issues/143 some thoughts have pop up for me regarding what we expect from the event holder in logstash. I want to open this issue so we can reach a common viewpoint on what we expect from it and also might generate a nice source of knowledge to be all in the same page for expectations when handing data into the event.

First of all what I've seen:

```bash
skywalker% cd logstash-2.2.0
skywalker% ls
CHANGELOG.md  CONTRIBUTORS  Gemfile  Gemfile.jruby-1.9.lock  LICENSE  NOTICE.TXT  bin  lib  vendor
skywalker% head Gemfile
# This is a Logstash generated Gemfile.
# If you modify this file manually all comments and formatting will be lost.

source "https://rubygems.org"
gem "logstash-core", "2.2.0"
gem "logstash-core-event", "2.2.0"
gem "file-dependencies", "0.1.6"
gem "ci_reporter_rspec", "1.0.0", :group => :development
gem "simplecov", :group => :development
gem "coveralls", :group => :development
skywalker% ./bin/logstash irb
irb(main):002:0> require "logstash/event"
=> true
irb(main):003:0> event = LogStash::Event.new({ "field" => "value".force_encoding("ISO-8859-1")})
=> #<LogStash::Event:0x6a56b1dc @metadata_accessors=#<LogStash::Util::Accessors:0x258fe6fa @store={}, @lut={}>, @cancelled=false, @data={"field"=>"value", "@version"=>"1", "@timestamp"=>"2016-06-30T09:28:39.300Z"}, @metadata={}, @accessors=#<LogStash::Util::Accessors:0x2e1c0089 @store={"field"=>"value", "@version"=>"1", "@timestamp"=>"2016-06-30T09:28:39.300Z"}, @lut={}>>
irb(main):004:0> event["field"]
=> "value"
irb(main):005:0> event["field"].encoding
=> #<Encoding:ISO-8859-1>
irb(main):006:0> 
```

```
skywalker% cd logstash-5.0.0-alpha3
skywalker% ls
CHANGELOG.md  CONTRIBUTORS  Gemfile  Gemfile.jruby-1.9.lock  LICENSE  NOTICE.TXT  bin  config  lib  vendor
skywalker% head Gemfile
# This is a Logstash generated Gemfile.
# If you modify this file manually all comments and formatting will be lost.

source "https://rubygems.org"
gem "logstash-core", "5.0.0.alpha3"
gem "logstash-core-event-java", "5.0.0.alpha3"
gem "logstash-core-plugin-api", "2.1.6"
gem "file-dependencies", "0.1.6"
gem "ci_reporter_rspec", "1.0.0", :group => :development
gem "simplecov", :group => :development
skywalker% ./bin/logstash irb
--- jar coordinate com.fasterxml.jackson.core:jackson-annotations already loaded with version 2.7.1 - omit version 2.7.0
--- jar coordinate com.fasterxml.jackson.core:jackson-databind already loaded with version 2.7.1 - omit version 2.7.1-1
ERROR: too many arguments

See: 'bin/logstash --help'
skywalker% ./bin/logstash --interactive=irb
--- jar coordinate com.fasterxml.jackson.core:jackson-annotations already loaded with version 2.7.1 - omit version 2.7.0
--- jar coordinate com.fasterxml.jackson.core:jackson-databind already loaded with version 2.7.1 - omit version 2.7.1-1
irb(main):001:0> require "logstash/event"
=> false
irb(main):002:0> event = LogStash::Event.new({ "field" => "value".force_encoding("ISO-8859-1")})
=> 2016-06-30T09:30:51.171Z %{host} %{message}
irb(main):004:0> event.get("field")
=> "value"
irb(main):005:0> event.get("field").encoding
=> #<Encoding:UTF-8>
irb(main):006:0> 
```

as you can see in the previous examples, this shows how for the ruby event the string encoding is 
preserved, for the java event is handled as `UTF-8`, keep in mind both initial strings where `ISO-8859-1`.

I've a few open questions regarding this that would be nice to see what do you think.

Until the introduction of the java event, the encoding was handled by the codecs through the [charset utility class](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/util/charset.rb), who basically try it best to encode strings to UTF-8. So basically the ruby event is a placeholder, what you put in is what you get out, responsability of the class client to do the encoding transformations.

With the introduction of the java event this behaviour changed now, if we put an non UTF-8 string it will be transformed to UTF-8 OOTB.

As I understand is in the benefit of the pipeline filters to have data transformed to UTF-8, for example for json, however, here are my open thoughts:

* Do we aim to keep the old ruby event behaviour with the java event?
* By opening the door to have non UTF-8 string encoding, are we having some filters fail, who might be expecting this format? how should they fail?
* Is the transformation to UTF-8 going to cause problem when people might like to keep their original encoding? does this makes sense?
* If as happens now, all strings are converted to `UTF-8` in the java event (like now), are the codecs that do charset transformations need anymore?

**Comment by colinsurprenant** [comment]
>  So basically the ruby event is a placeholder, what you put in is what you get out, responsability of the class client to do the encoding transformations.

Yes and the convention is to have all values always UTF-8 encoded when setting them into the Event. This is not enforced in the Event code, but it is a convention. At some point it was decided that doing a systematic encoding validation check for all values set in the Event was going to be too costly. 

> With the introduction of the java event this behaviour changed now, if we put an non UTF-8 string it will be transformed to UTF-8 OOTB.

This is an implementation specific side effect, not the result of a new explicit rule or validation put in place.

> Do we aim to keep the old ruby event behaviour with the java event?

This is irrelevant. The first example you provided is **not supposed** to happen. This is the plugin job & responsibility to make sure every values it sets in the event is UTF-8. 

> By opening the door to have non UTF-8 string encoding, are we having some filters fail, who might be expecting this format? how should they fail?

We are not opening the door. We are just not validating the encoding on very event value set. 

> Is the transformation to UTF-8 going to cause problem when people might like to keep their original encoding? does this makes sense?

Within logstash it is essential to rely on a single encoding (UTF-8) and this is what the codecs `charset` option is used-for on the input side. For the output side, I am not sure there are many practical use-cases of using other encodings, I haven't seen any user request for this. Nothing would prevent us to use the same charset encoding on the output+codec side. For a non-codec specific output, it is completely free to transcode to any target charset it need for its specific protocol/format.

> If as happens now, all strings are converted to UTF-8 in the java event (like now), are the codecs that do charset transformations need anymore?

Absolutely. A **proper** transcoding need to be aware of the correct source and target charsets. Anything else is speculations on the resulting UTF-8.

**Comment by colinsurprenant** [comment]
Are we through with this discussion?

**Comment by purbon** [comment]
Makes sense @colinsurprenant, I have however another doubt here, as we leave the responsibility for encoding to the input side, for situations where the input fails to provide proper encoding, do you think we could work on providing more specific errors on specific encoding dependent filters? my guess here is, for example:

* jdbc input fails to set proper encoding,
* later in the config, we find a filter, for example xml, that needs proper encoding, so it complains.

should we workout on this plugins and raise more explicit exception when this kind of errors happen?

what do you think?

**Comment by purbon** [comment]
As this is the standard for us on how to deal with encoding, should be document this somewhere? might be in some developer directions document?

**Comment by colinsurprenant** [comment]
@purbon agree on better documentation.

For the check, this is tricky, as I said earlier, this is a decision we took: not to do systematic encoding checks. This is why we had the systematic encoding check ONLY in specs, where these errors should normally surface but in some cases, like JDBC, where this is use-case & environment specific, specs may not catch all potential & future problems with that. 

Maybe we could explore if, globally, encoding exceptions could be better trapped and reported, but this has not been such a huge problem for users for a while so I am not sure about the importance vs urgency of this issue.

**Comment by colinsurprenant** [comment]
another idea, maybe we could introduce systematic encoding verification through a config/command-line flag? this way, upon encoding errors, one could enable it to better diagnose the source of the problem? this flag could also be used by specs instead of the funky monkey patching?
